src/d3a/models/state.py

```diff
@@ -43,131 +46,247 @@
+        return self._energy_requirement_Wh.get(time_slot, default_value)
+
+    def set_desired_energy(self, energy, time_slot, overwrite=False):
+        if overwrite is False and time_slot in self._energy_requirement_Wh:
+            return
+        self._energy_requirement_Wh[time_slot] = energy
+        self._desired_energy_Wh[time_slot] = energy
+
+    def update_total_demanded_energy(self, time_slot):
+        self._total_energy_demanded_Wh += self._desired_energy_Wh.get(time_slot, 0.)
+
+    def can_buy_more_energy(self, time_slot):
+        if time_slot not in self._energy_requirement_Wh:
+            return False
+
+        return self._energy_requirement_Wh[time_slot] > FLOATING_POINT_TOLERANCE
+
+    def calculate_energy_to_accept(self, offer_energy_Wh: float, time_slot: DateTime) -> float:
+        """Return the amount of energy that can be accepted in a specific market slot.
+
+        The acceptable energy can't be more than the total energy required in that slot.
+        """
+        return min(offer_energy_Wh, self._energy_requirement_Wh[time_slot])
+
+    def decrement_energy_requirement(
+            self, purchased_energy_Wh: float, time_slot: DateTime, area_name: str) -> None:
+        """Decrease the energy required by the device in a specific market slot."""
+        self._energy_requirement_Wh[time_slot] -= purchased_energy_Wh
+        assert self._energy_requirement_Wh[time_slot] >= -FLOATING_POINT_TOLERANCE, (
+            f"Energy requirement for device {area_name} fell below zero "
+            f"({self._energy_requirement_Wh[time_slot]}).")
+
+    def delete_past_state_values(self, current_time_slot):
+        """Delete data regarding energy consumption for past market slots."""
+        to_delete = []
+        for market_slot in self._energy_requirement_Wh.keys():
+            if market_slot < current_time_slot:
+                to_delete.append(market_slot)
+
+        for market_slot in to_delete:
+            self._energy_requirement_Wh.pop(market_slot, None)
+            self._desired_energy_Wh.pop(market_slot, None)
+
+    def get_desired_energy_Wh(self, time_slot, default_value=0.0):
+        """Return the expected consumed energy at a specific market slot."""
+        return self._desired_energy_Wh.get(time_slot, default_value)
+
+
+class ProductionState(StateInterface):
+    """State for devices that can produce energy."""
+
     def __init__(self):
+        super().__init__()
         self._available_energy_kWh = {}
         self._energy_production_forecast_kWh = {}
 
-    def get_energy_production_forecast_kWh(self, time_slot, default_value=None):
-        assert self._energy_production_forecast_kWh[time_slot] >= -FLOATING_POINT_TOLERANCE
-        if default_value is not None:
-            return self._energy_production_forecast_kWh.get(time_slot, default_value)
-        return self._energy_production_forecast_kWh[time_slot]
+    def get_state(self) -> Dict:
```

Comment:
I see that this method is mostly repeating itself for the strategy classes, but unfortunately I do not have a good way off the top of my head to make it more DRY. Maybe when I stumble upon it in the future I'll also try to fight it to remove the repeating code